{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4048c0c3-7a37-4173-be60-f5eed24e38eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "Q1. Explain why we have to use the Exception class while creating a Custom Exception.\n",
    "Note: Here Exception class refers to the base class for all the exceptions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2120f45a-c954-41f1-90d9-a4ba98ff4cc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "In Python, the Exception class is the base class for all built-in and custom exceptions. When we create a custom exception,\n",
    "we typically inherit from the Exception class to create a new exception type that has specific behaviors or properties.\n",
    "\n",
    "There are several reasons why we use the Exception class as a base class when creating custom exceptions:\n",
    "\n",
    "0. Inheritance: By inheriting from the Exception class, our custom exception automatically inherits all the properties and\n",
    "   behaviors of the Exception class. This includes the ability to add custom error messages, tracebacks, and \n",
    "   other information that can help debug issues in our code.\n",
    "\n",
    "1. Compatibility: By inheriting from the Exception class, our custom exception is compatible with all existing code that expects to handle exceptions \n",
    "   of type Exception. This allows us to use our custom exception seamlessly in any code that handles exceptions.\n",
    "\n",
    "2. Standardization: By using the Exception class as a base class, we adhere to the standard practice for creating exceptions in Python. This makes our\n",
    "   code more readable and easier to maintain for other developers who may be working on the same codebase.\n",
    "\n",
    "3. Customization: By inheriting from the Exception class, we can customize our custom exception to suit our specific needs. We can add new properties \n",
    "   or methods that are specific to our custom exception, and override existing methods to provide custom behavior when the exception is raised.\n",
    "\n",
    "In summary, using the Exception class as a base class when creating custom exceptions provides many benefits, including inheritance, compatibility, \n",
    "standardization, and customization. It is considered best practice to use the Exception class as a base class when creating custom exceptions in\n",
    "Python."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9dda65a0-27d1-4942-a5a0-8a7c3ad0812e",
   "metadata": {},
   "outputs": [],
   "source": [
    "Q2. Write a python program to print Python Exception Hierarchy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2c815a97-cab7-45f6-affc-4cd3f3e8e4bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Python Exception Hierarchy:\n",
      "--------------------------\n",
      "BaseException\n",
      "AbortThread\n",
      "CancelledError\n",
      "KeyboardInterrupt\n",
      "SystemExit\n",
      "GeneratorExit\n",
      "Exception\n",
      "InvalidTypeInArgsException\n",
      "UnableToResolveVariableException\n",
      "_Error\n",
      "ResolutionError\n",
      "UnknownExtra\n",
      "DistributionNotFound\n",
      "VersionConflict\n",
      "ContextualVersionConflict\n",
      "RecursiveGrammarException\n",
      "ParseBaseException\n",
      "ParseFatalException\n",
      "ParseSyntaxException\n",
      "ParseException\n",
      "Error\n",
      "Fault\n",
      "ResponseError\n",
      "ProtocolError\n",
      "ExpatError\n",
      "DebuggerInitializationError\n",
      "QueueFull\n",
      "QueueEmpty\n",
      "_Ipv6UnsupportedError\n",
      "Error\n",
      "TimeoutExpired\n",
      "AccessDenied\n",
      "NoSuchProcess\n",
      "ZombieProcess\n",
      "KillEmbedded\n",
      "InteractivelyDefined\n",
      "HTTPException\n",
      "LineTooLong\n",
      "BadStatusLine\n",
      "RemoteDisconnected\n",
      "ImproperConnectionState\n",
      "ResponseNotReady\n",
      "CannotSendHeader\n",
      "CannotSendRequest\n",
      "IncompleteRead\n",
      "UnimplementedFileMode\n",
      "UnknownTransferEncoding\n",
      "UnknownProtocol\n",
      "InvalidURL\n",
      "NotConnected\n",
      "Error\n",
      "MessageError\n",
      "CharsetError\n",
      "MultipartConversionError\n",
      "MessageParseError\n",
      "BoundaryError\n",
      "HeaderParseError\n",
      "InvalidPythonEnvironment\n",
      "OnErrorLeaf\n",
      "_JediError\n",
      "RefactoringError\n",
      "WrongVersion\n",
      "InternalError\n",
      "ParamIssue\n",
      "HasNoContext\n",
      "UncaughtAttributeError\n",
      "SimpleGetItemNotFound\n",
      "_PositionUpdatingFinished\n",
      "InternalParseError\n",
      "ParserSyntaxError\n",
      "HeightIsUnknownError\n",
      "InvalidLayoutError\n",
      "_Retry\n",
      "EditReadOnlyBuffer\n",
      "ValidationError\n",
      "DOMException\n",
      "ValidationErr\n",
      "InvalidAccessErr\n",
      "NamespaceErr\n",
      "InvalidModificationErr\n",
      "SyntaxErr\n",
      "InvalidStateErr\n",
      "InuseAttributeErr\n",
      "NotSupportedErr\n",
      "NotFoundErr\n",
      "NoModificationAllowedErr\n",
      "NoDataAllowedErr\n",
      "InvalidCharacterErr\n",
      "WrongDocumentErr\n",
      "HierarchyRequestErr\n",
      "DomstringSizeErr\n",
      "IndexSizeErr\n",
      "SpaceInInput\n",
      "Warning\n",
      "Error\n",
      "DatabaseError\n",
      "NotSupportedError\n",
      "DataError\n",
      "IntegrityError\n",
      "ProgrammingError\n",
      "OperationalError\n",
      "InternalError\n",
      "InterfaceError\n",
      "AliasError\n",
      "InvalidAliasError\n",
      "PrefilterError\n",
      "ErrorToken\n",
      "GetoptError\n",
      "InputRejected\n",
      "IPythonCoreError\n",
      "StdinNotImplementedError\n",
      "UsageError\n",
      "TryNext\n",
      "ProfileDirError\n",
      "HomeDirError\n",
      "FindCmdError\n",
      "PtyProcessError\n",
      "ExceptionPexpect\n",
      "TIMEOUT\n",
      "EOF\n",
      "Restart\n",
      "BdbQuit\n",
      "OptionError\n",
      "CannotEval\n",
      "NotOneValueFound\n",
      "ErrorDuringImport\n",
      "DuplicateKernelError\n",
      "NoSuchEntryPoint\n",
      "BadEntryPoint\n",
      "LargeZipFile\n",
      "BadZipFile\n",
      "NoIPAddresses\n",
      "Error\n",
      "ParsingError\n",
      "MissingSectionHeaderError\n",
      "InterpolationError\n",
      "InterpolationDepthError\n",
      "InterpolationSyntaxError\n",
      "InterpolationMissingOptionError\n",
      "NoOptionError\n",
      "DuplicateOptionError\n",
      "DuplicateSectionError\n",
      "NoSectionError\n",
      "_GiveupOnFastCopy\n",
      "RegistryError\n",
      "LZMAError\n",
      "error\n",
      "InvalidPortNumber\n",
      "Return\n",
      "ReturnValueIgnoredError\n",
      "BadYieldError\n",
      "LeakedCallbackError\n",
      "UnknownKeyError\n",
      "KeyReuseError\n",
      "ReturnValueIgnoredError\n",
      "error\n",
      "TimeoutError\n",
      "error\n",
      "ApplicationError\n",
      "ConfigurableError\n",
      "MultipleInstanceError\n",
      "ConfigError\n",
      "ConfigFileNotFound\n",
      "ConfigLoaderError\n",
      "ArgumentError\n",
      "ArgumentTypeError\n",
      "ArgumentError\n",
      "_Stop\n",
      "PickleError\n",
      "UnpicklingError\n",
      "PicklingError\n",
      "ZMQBaseError\n",
      "NotDone\n",
      "ZMQBindError\n",
      "ZMQError\n",
      "InterruptedSystemCall\n",
      "Again\n",
      "ContextTerminated\n",
      "ArgumentError\n",
      "Full\n",
      "Empty\n",
      "QueueFull\n",
      "QueueEmpty\n",
      "LimitOverrunError\n",
      "InvalidStateError\n",
      "TimeoutError\n",
      "Incomplete\n",
      "error\n",
      "_GiveupOnSendfile\n",
      "Error\n",
      "InvalidStateError\n",
      "TimeoutError\n",
      "CancelledError\n",
      "Error\n",
      "TraitError\n",
      "EndOfBlock\n",
      "ClassFoundException\n",
      "StopTokenizing\n",
      "TokenError\n",
      "SubprocessError\n",
      "TimeoutExpired\n",
      "CalledProcessError\n",
      "Error\n",
      "Verbose\n",
      "error\n",
      "_Error\n",
      "_OptionError\n",
      "Warning\n",
      "PkgResourcesDeprecationWarning\n",
      "DeprecatedTzFormatWarning\n",
      "ResourceWarning\n",
      "BytesWarning\n",
      "UnicodeWarning\n",
      "ImportWarning\n",
      "FutureWarning\n",
      "ProvisionalCompleterWarning\n",
      "RuntimeWarning\n",
      "PEP440Warning\n",
      "UnknownTimezoneWarning\n",
      "ProactorSelectorThreadWarning\n",
      "SyntaxWarning\n",
      "PendingDeprecationWarning\n",
      "DeprecationWarning\n",
      "ProvisionalWarning\n",
      "EncodingWarning\n",
      "UserWarning\n",
      "FormatterWarning\n",
      "GetPassWarning\n",
      "BufferError\n",
      "MemoryError\n",
      "ReferenceError\n",
      "SystemError\n",
      "CodecRegistryError\n",
      "ArithmeticError\n",
      "DecimalException\n",
      "InvalidOperation\n",
      "InvalidContext\n",
      "DivisionUndefined\n",
      "DivisionImpossible\n",
      "ConversionSyntax\n",
      "FloatOperation\n",
      "DivisionByZero\n",
      "Subnormal\n",
      "Underflow\n",
      "Inexact\n",
      "Overflow\n",
      "Underflow\n",
      "Rounded\n",
      "Overflow\n",
      "Underflow\n",
      "Clamped\n",
      "ZeroDivisionError\n",
      "DivisionUndefined\n",
      "DivisionByZero\n",
      "OverflowError\n",
      "FloatingPointError\n",
      "AssertionError\n",
      "ValueError\n",
      "InvalidVersion\n",
      "InvalidRequirement\n",
      "RequirementParseError\n",
      "UndefinedEnvironmentName\n",
      "UndefinedComparison\n",
      "InvalidMarker\n",
      "InvalidSpecifier\n",
      "InvalidSdistFilename\n",
      "InvalidWheelFilename\n",
      "_InvalidELFFileHeader\n",
      "InvalidVersion\n",
      "UnequalIterablesError\n",
      "InvalidFileException\n",
      "MacroToEdit\n",
      "MessageDefect\n",
      "HeaderDefect\n",
      "InvalidDateDefect\n",
      "NonASCIILocalPartDefect\n",
      "ObsoleteHeaderDefect\n",
      "NonPrintableDefect\n",
      "HeaderMissingRequiredValue\n",
      "InvalidHeaderDefect\n",
      "InvalidBase64LengthDefect\n",
      "InvalidBase64CharactersDefect\n",
      "InvalidBase64PaddingDefect\n",
      "UndecodableBytesDefect\n",
      "InvalidMultipartContentTransferEncodingDefect\n",
      "MultipartInvariantViolationDefect\n",
      "MissingHeaderBodySeparatorDefect\n",
      "MisplacedEnvelopeHeaderDefect\n",
      "FirstHeaderLineIsContinuationDefect\n",
      "CloseBoundaryNotFoundDefect\n",
      "StartBoundaryNotFoundDefect\n",
      "NoBoundaryInMultipartDefect\n",
      "ClipboardEmpty\n",
      "ClassNotFound\n",
      "ParserError\n",
      "IllegalWeekdayError\n",
      "IllegalMonthError\n",
      "UnsupportedDigestmodError\n",
      "Error\n",
      "SSLCertVerificationError\n",
      "JSONDecodeError\n",
      "UnsupportedOperation\n",
      "UnicodeError\n",
      "UnicodeTranslateError\n",
      "UnicodeDecodeError\n",
      "UnicodeEncodeError\n",
      "LookupError\n",
      "CodecRegistryError\n",
      "KeyError\n",
      "UnknownBackend\n",
      "NoSuchKernel\n",
      "IndexError\n",
      "SyntaxError\n",
      "IndentationError\n",
      "TabError\n",
      "AttributeError\n",
      "FrozenInstanceError\n",
      "NameError\n",
      "UnboundLocalError\n",
      "RuntimeError\n",
      "VariableError\n",
      "ExtractionError\n",
      "SendfileNotAvailableError\n",
      "BrokenExecutor\n",
      "BrokenThreadPool\n",
      "BrokenBarrierError\n",
      "_DeadlockError\n",
      "NotImplementedError\n",
      "StdinNotImplementedError\n",
      "ZMQVersionError\n",
      "RecursionError\n",
      "EOFError\n",
      "IncompleteReadError\n",
      "OSError\n",
      "BadGzipFile\n",
      "URLError\n",
      "ContentTooShortError\n",
      "HTTPError\n",
      "ReadError\n",
      "ExecError\n",
      "SpecialFileError\n",
      "Error\n",
      "SameFileError\n",
      "SSLError\n",
      "SSLEOFError\n",
      "SSLSyscallError\n",
      "SSLWantReadError\n",
      "SSLWantWriteError\n",
      "SSLZeroReturnError\n",
      "SSLCertVerificationError\n",
      "gaierror\n",
      "herror\n",
      "itimer_error\n",
      "UnsupportedOperation\n",
      "TimeoutError\n",
      "ProcessLookupError\n",
      "PermissionError\n",
      "InterruptedError\n",
      "InterruptedSystemCall\n",
      "NotADirectoryError\n",
      "IsADirectoryError\n",
      "FileNotFoundError\n",
      "FileExistsError\n",
      "ChildProcessError\n",
      "BlockingIOError\n",
      "ConnectionError\n",
      "ConnectionResetError\n",
      "RemoteDisconnected\n",
      "ConnectionRefusedError\n",
      "ConnectionAbortedError\n",
      "BrokenPipeError\n",
      "ImportError\n",
      "ZipImportError\n",
      "ModuleNotFoundError\n",
      "StopIteration\n",
      "StopAsyncIteration\n",
      "TypeError\n",
      "MultipartConversionError\n",
      "FloatOperation\n"
     ]
    }
   ],
   "source": [
    "def print_exception_hierarchy():\n",
    "    \"\"\"\n",
    "    This function prints the Python Exception Hierarchy.\n",
    "    \"\"\"\n",
    "    print(\"Python Exception Hierarchy:\")\n",
    "    print(\"--------------------------\")\n",
    "    base_exception = BaseException\n",
    "    exceptions = [base_exception]\n",
    "    while exceptions:\n",
    "        exception = exceptions.pop()\n",
    "        print(exception.__name__)\n",
    "        for subclass in exception.__subclasses__():\n",
    "            exceptions.append(subclass)\n",
    "            \n",
    "print_exception_hierarchy()\n",
    "            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ec02a4c-a2cf-40e0-9350-6426fceb16c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "Q3. What errors are defined in the ArithmeticError class? Explain any two with an example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6efb1551-0066-40f4-a465-225c568173b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "In Python, the ArithmeticError class is a base class for all exceptions that occur during arithmetic operations.\n",
    "It is a subclass of the Exception class and a superclass of specific arithmetic exceptions like ZeroDivisionError, OverflowError,\n",
    "and FloatingPointError. The ArithmeticError class itself does not define any specific error types, but rather serves\n",
    "as a general category for arithmetic-related exceptions.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0c9dcc10-5a4b-4a5d-8d6b-31d9e641c384",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "division by zero\n"
     ]
    }
   ],
   "source": [
    "## 1. ZeroDivisionError\n",
    "try:\n",
    "    15/0\n",
    "except ZeroDivisionError as e:\n",
    "    print(e)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "04eb7f38-05dd-453d-8e10-0f05bff4974b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 2.OverflowError\n",
    "import sys\n",
    "a = sys.maxsize\n",
    "b = a + 1  # OverflowError: Python int too large to convert to C long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad8f0d4b-104f-4404-88ca-025b101223b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "Q4. Why LookupError class is used? Explain with an example KeyError and IndexError."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45933f1c-5d7d-4de5-99c2-fccf996cc2ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "The LookupError class is a base class for all exceptions that occur when a key or index lookup fails. It is a subclass of the Exception class and\n",
    "is itself a superclass of specific lookup exceptions like KeyError and IndexError. The LookupError class is used as a general category for \n",
    "lookup-related exceptions, and it is often used to catch and handle errors that occur during dictionary lookups or list indexing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "900f3f4b-02a1-45ca-98f2-4d7ecd5ec489",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'pear'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[8], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m my_dict \u001b[38;5;241m=\u001b[39m {\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mapple\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;241m1\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mbanana\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;241m2\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124morange\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;241m3\u001b[39m}\n\u001b[0;32m----> 2\u001b[0m value \u001b[38;5;241m=\u001b[39m \u001b[43mmy_dict\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mpear\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m  \u001b[38;5;66;03m# KeyError: 'pear'\u001b[39;00m\n\u001b[1;32m      4\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m      5\u001b[0m     d \u001b[38;5;241m=\u001b[39m {\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mkey\u001b[39m\u001b[38;5;124m\"\u001b[39m:\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpradeep\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;241m1\u001b[39m :[\u001b[38;5;241m1\u001b[39m,\u001b[38;5;241m2\u001b[39m,\u001b[38;5;241m34\u001b[39m,\u001b[38;5;241m5\u001b[39m]}\n",
      "\u001b[0;31mKeyError\u001b[0m: 'pear'"
     ]
    }
   ],
   "source": [
    "## example 1\n",
    "my_dict = {'apple': 1, 'banana': 2, 'orange': 3}\n",
    "value = my_dict['pear']  # KeyError: 'pear'\n",
    "\n",
    "\n",
    "try:\n",
    "    d = {\"key\":\"pradeep\", 1 :[1,2,34,5]}\n",
    "    print(d[\"key2\"])\n",
    "except KeyError as e:\n",
    "    print(e)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e9a83258-571d-4276-9406-705269897a37",
   "metadata": {},
   "outputs": [
    {
     "ename": "IndexError",
     "evalue": "list index out of range",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[7], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m my_list \u001b[38;5;241m=\u001b[39m [\u001b[38;5;241m1\u001b[39m, \u001b[38;5;241m2\u001b[39m, \u001b[38;5;241m3\u001b[39m]\n\u001b[0;32m----> 2\u001b[0m value \u001b[38;5;241m=\u001b[39m \u001b[43mmy_list\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m3\u001b[39;49m\u001b[43m]\u001b[49m  \u001b[38;5;66;03m# IndexError: list index out of range\u001b[39;00m\n",
      "\u001b[0;31mIndexError\u001b[0m: list index out of range"
     ]
    }
   ],
   "source": [
    "my_list = [1, 2, 3]\n",
    "value = my_list[3]  # IndexError: list index out of range"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c045df42-7c9b-4d57-9e6b-cd6b241beb8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "Q5. Explain ImportError. What is ModuleNotFoundError?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75278ec2-eec1-4116-b09a-5e5f4bac837d",
   "metadata": {},
   "outputs": [],
   "source": [
    "ImportError is a built-in exception in Python that is raised when a module or package cannot be imported. It is a subclass of the Exception \n",
    "class and is often used to catch and handle import-related errors.\n",
    "\n",
    "There are several reasons why an ImportError might be raised. For example, it could be because the module or package does not exist in the \n",
    "specified location, the file has incorrect permissions, or the module is not installed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "cf728fc8-1069-4be3-95d7-d5661ac4dace",
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "logging.basicConfig(filename  = \"error.log\",level = logging.ERROR)\n",
    "try:\n",
    "    import my_module\n",
    "except ModuleNotFoundError as e :\n",
    "    logging.error(\"The module 'my_module' could not be found. {}\".format(e))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79054279-ede2-4d43-b7b6-44cbce2cae82",
   "metadata": {},
   "outputs": [],
   "source": [
    "Q6. List down some best practices for exception handling in python. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "816234fc-6074-43dc-9aeb-8f5755b0a468",
   "metadata": {},
   "outputs": [],
   "source": [
    "Use the appropriate exception type\n",
    "Don't catch all exceptions\n",
    "Use finally to clean up resources\n",
    "Avoid raising generic exceptions\n",
    "Document exception handling\n",
    "cleanup all the resources"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
